packages/ic-http-certification/src/http/http_request.rs

```diff
@@ -1,39 +1,766 @@
 #[derive(Clone, Debug, CandidType, Deserialize, PartialEq, Eq)]
-pub struct HttpRequest {
+pub struct HttpRequest<'a> {
     /// HTTP request method.
-    pub method: String,
-    /// Request URL.
-    pub url: String,
+    method: Cow<'a, str>,
+
+    /// HTTP request URL.
+    url: Cow<'a, str>,
+
+    /// HTTP request headers.
+    headers: Vec<HeaderField>,
+
+    /// HTTP request body as an array of bytes.
+    body: Cow<'a, [u8]>,
+
+    /// The max response verification version to use in the response's
+    /// certificate.
+    certificate_version: Option<u16>,
+}
+
+impl<'a> HttpRequest<'a> {
+    /// Creates a new [HttpRequestBuilder] initialized with a GET method and
+    /// the given URL.
+    ///
+    /// This method returns an instance of [HttpRequestBuilder] which can be
+    /// used to create an [HttpRequest].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_http_certification::HttpRequest;
+    ///
+    /// let request = HttpRequest::get("/").build();
+    ///
+    /// assert_eq!(request.method(), "GET");
+    /// ```
+    pub fn get(url: impl Into<Cow<'a, str>>) -> HttpRequestBuilder<'a> {
+        HttpRequestBuilder::new().with_method("GET").with_url(url)
+    }
+
+    /// Creates a new [HttpRequestBuilder] initialized with a POST method and
+    /// the given URL.
+    ///
+    /// This method returns an instance of [HttpRequestBuilder] which can be
+    /// used to create an [HttpRequest].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_http_certification::HttpRequest;
+    ///
+    /// let request = HttpRequest::post("/").build();
+    ///
+    /// assert_eq!(request.method(), "POST");
+    /// ```
+    pub fn post(url: impl Into<Cow<'a, str>>) -> HttpRequestBuilder<'a> {
+        HttpRequestBuilder::new().with_method("POST").with_url(url)
+    }
+
+    /// Creates a new [HttpRequestBuilder] initialized with a PUT method and
+    /// the given URL.
+    ///
+    /// This method returns an instance of [HttpRequestBuilder] which can be
+    /// used to create an [HttpRequest].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_http_certification::HttpRequest;
+    ///
+    /// let request = HttpRequest::put("/").build();
+    ///
+    /// assert_eq!(request.method(), "PUT");
+    /// ```
+    pub fn put(url: impl Into<Cow<'a, str>>) -> HttpRequestBuilder<'a> {
+        HttpRequestBuilder::new().with_method("PUT").with_url(url)
+    }
+
+    /// Creates a new [HttpRequestBuilder] initialized with a PATCH method and
+    /// the given URL.
+    ///
+    /// This method returns an instance of [HttpRequestBuilder] which can be
+    /// used to create an [HttpRequest].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_http_certification::HttpRequest;
+    ///
+    /// let request = HttpRequest::patch("/").build();
+    ///
+    /// assert_eq!(request.method(), "PATCH");
+    /// ```
+    pub fn patch(url: impl Into<Cow<'a, str>>) -> HttpRequestBuilder<'a> {
```

Comment:
Not sure whether we should have the convenience helpers also for the less-used methods, as they blow-up the API considerably... Maybe `get()`, `post()`, and `put()` would be sufficient? Wdyt?

Reply:
I've personally used GET, POST, PUT, PATCH, DELETE quite a lot. So I think adding PATCH and DELETE to that list would be nice.

OPTIONS is normally handled automatically by web server frameworks. I just realized that the HEAD method is missing anyway, but that method along with TRACE and CONNECT are methods I've personally never used.

So what do you think of removing OPTIONS, TRACE and CONNECT?

Reply:
Agreed, covering also other methods that are used frequently makes sense, and yeah, I'd suggest removing the ones that are rarely used.